src/hotspot/share/runtime/deoptimization.cpp

```diff
@@ -440,8 +440,15 @@ static void restore_eliminated_locks(JavaThread* thread, GrowableArray<compiledV
     }
   }
 #ifdef ASSERT
-  if (LockingMode == LM_LIGHTWEIGHT && !realloc_failures) {
-    deoptee_thread->lock_stack().verify_consistent_lock_order(lock_order, exec_mode != Deoptimization::Unpack_none);
+  if (LockingMode == LM_LIGHTWEIGHT && !realloc_failures && lock_order.is_nonempty()) {
+    const int bci = chunk->first()->bci();
+    const Bytecodes::Code bc = chunk->first()->method()->code_at(bci);
+    const bool is_syncronized_entry = chunk->first()->method()->is_synchronized() &&
+                                      chunk->first()->raw_bci() == SynchronizationEntryBCI;
+    // If deoptimizing from monitorenter bytecode we maybe in transitional state. Skip verification.
+    if (!is_syncronized_entry && bc != Bytecodes::Code::_monitorenter) {
+      deoptee_thread->lock_stack().verify_consistent_lock_order(lock_order, exec_mode != Deoptimization::Unpack_none);
+    }
```

Comment:
The above checks would also hit the follow false positives:
1. deopt in counter overflow in prologue, not in monitorenter
2. monitorenter at bci 0 when raw_bci is -1 (assuming it got past the verifier)
but seems mostly harmless to skip checks in those cases.

Reply:
I thought the original check was fine. Could you elaborate on these 2 cases, I didn't really get them.

Reply:
Deoptimization is already expensive, and this edge case is rare, so I think it would be better to compute the actual previous bytecode here, and not use bci - 1.

Reply:
Other debug checks in deoptimization compute oop maps, which have to iterate all the bytecodes, so doing it here also wouldn't be so bad.  Or how about not checking bytecodes and instead checking a flag on JavaThread that says we are in monitor enter native code?

Reply:
> Deoptimization is already expensive, and this edge case is rare, so I think it would be better to compute the actual previous bytecode here, and not use bci - 1.

Alright I will give that a go then, unless we think the second option is more appropriate. 

> Or how about not checking bytecodes and instead checking a flag on JavaThread that says we are in monitor enter native code?

What state would that be? Any current state we setup is after transitioning to VM which seems to late. I guess it would be possible to instrument `SharedRuntime::monitor_enter_helper` with some thread local state we can check. But the interpreter has its own entry point. But I guess we would never reach this point for the interpreter, as there is no such thing as deoptimizing or unpacking interpreted frames. 


What is the feeling here? What would be more appropriate? Adding some thread local debug state that we set before transitioning to VM in `SharedRuntime::monitor_enter_helper`  seems like the most precise solution, but we need to be sure that there are no earlier safepoint polls before this point within the execution of the monitorenter bytecode.

Reply:
I see there is `_pending_monitorenter` but this would only handle synchronized method entry.

Reply:
I like the idea of a flag better, because it is foolproof.  Why can't we set it in ObjectSynchronizer::enter?  I don't think it matters if there is a safepoint check before that, because the lock stack is still consistent at that point.

Reply:
(Currently) the lock stack is always consistent at safepoints w.r.t. what is actually locked. However the lock stack may not be consistent with the most recent lock returned by the leaf `compiledVFrame::monitors()`. 

But you are correct that it can probably be moved to `ObjectSynchronizer::enter` there are no safepoint polls between `SharedRuntime::monitor_enter_helper` and that point. Similarly there are no safepoints polls in the runtime until after `set_current_pending_monitor` is called. 

So with these following assumptions.
1. LockStack is consistent at safepoints w.r.t. locked monitors
2. No safepoint polls exist from the point that compiledVFrame::monitors() starts returning the monitorinfo for the currently executing monitorenter until either it calls into the runtime or finishes locking.

I do not believe 1. is likely to ever change. But I have limited understanding of the validity of 2. nor if it something that can change.

If both these assumptions are correct than simply skipping the verification when `deoptee->current_pending_monitor() != nullptr` would suffice.

Reply:
I believe those assumptions will always hold, but per separate discussions, let's go with what you have.